src/allmydata/storage_client_http.py

```diff
@@ -0,0 +1,204 @@
+                for share_num in result.allocated
+             }
+        )
+
+    def add_lease(self, storage_index, renew_secret, cancel_secret):
+        return self._client.add_lease(
+            storage_index, renew_secret, cancel_secret
+        )
+
+    @inlineCallbacks
+    def get_buckets(self, storage_index):
+        share_numbers = yield self._client.immutable_list_shares(
+            storage_index
+        )
+        returnValue({
+            share_num: _ClientV2BucketReader(
+                self._client, storage_index, share_num
+            )
+            for share_num in share_numbers
+        })
+
+    @inlineCallbacks
+    def slot_readv(self, storage_index, shares, readv):
+        reads = {}
+        for share_number in shares:
+            share_reads = reads[share_number] = []
+            for (offset, length) in readv:
+                d = self._client.mutable.read_share_chunk(
+                    storage_index, share_number, offset, length
+                )
+                share_reads.append(d)
+        result = {
+            share_number: [(yield d) for d in share_reads]
+            for (share_number, reads) in reads.items()
+        }
+        returnValue(result)
+
+    def slot_testv_and_readv_and_writev(
+            self,
+            storage_index,
+            secrets,
+            tw_vectors,
+            r_vector,
+    ):
+        we_secret, lr_secret, lc_secret = secrets
+        client_tw_vectors = {}
+        for share_num, (test_vector, data_vector, new_length) in tw_vectors.items():
+            assert new_length is not None, "Protocol in theory supports it, actual code seems not to"
+            client_test_vectors = [
+                TestVector(offset, size, TestVectorOperator[op], specimen)
+                for (offset, size, op, specimen) in test_vector
+            ]
+            client_write_vectors = [
+                WriteVector(offset, data) for (offset, data) in data_vector
+            ]
+            client_tw_vectors[share_num] = TestWriteVectors(
+                test_vectors=client_test_vectors,
+                write_vectors=client_write_vectors,
+                new_length=new_length
+            )
+        client_read_vectors = [
+            ReadVector(offset=offset, size=size)
+            for (offset, size) in r_vector
+        ]
+        client_result = yield self._client.mutable_read_test_write_chunk(
+            storage_index, we_secret, lr_secret, lc_secret, client_tw_vectors,
+            client_read_vectors,
+        )
+        returnValue((client_result.success, client_result.reads))
+
+    def advise_corrupt_share(self, share_type, storage_index, shnum, reason):
+        reason = str(reason, "utf-8", errors="backslashreplace")
+        if share_type == b"mutable":
+            advise = self._client.mutable_notify_shared_corrupted
+        elif share_type == b"immutable":
+            advise = self._client.immutable_notify_shared_corrupted
+        else:
```

Comment:
Reading this now I'm not sure what `share_type` is actually supposed to mean.  In this particular code what interpretation is being applied - but maybe I got it wrong in the GBS spec itself.

The client can say anything here, of course.  One might hope it is making its best effort to tell us the correct share type but it could always be in error somehow.

But ... so what, maybe?  The only thing the server does with the information is interpolate it into a plaintext string in a couple places.

I wonder if GBS should keep its two variations of this endpoint, though, or merge them into one with a `share_type` parameter, more closely mirroring the Foolscap API.

Maybe this isn't actionable here, but rather on some other ticket related to spec fixes.

Reply:
Having two endpoints seems nicer, insofar as there's already two namespaces in the URL for immutable vs. mutable?